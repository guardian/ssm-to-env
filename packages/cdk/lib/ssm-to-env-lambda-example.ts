import {
	GuDistributionBucketParameter,
	GuStack,
} from '@guardian/cdk/lib/constructs/core';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
import type { App } from 'aws-cdk-lib';
import { Code, LayerVersion, Runtime } from 'aws-cdk-lib/aws-lambda';
import { Bucket } from 'aws-cdk-lib/aws-s3';

interface SsmToEnvLambdaStackProps extends GuStackProps {
	app: string;
}

export class SsmToEnvLambdaExample extends GuStack {
	constructor(scope: App, id: string, props: SsmToEnvLambdaStackProps) {
		super(scope, id, props);

		const bucket = Bucket.fromBucketName(
			this,
			`${id}-bucket`,
			GuDistributionBucketParameter.getInstance(this).valueAsString,
		);

		const keyPrefix = `${this.stack}/${this.stage}/${props.app}`;

		const getSecretsLayer = new LayerVersion(this, 'get-secrets-layer', {
			code: Code.fromBucket(
				bucket,
				`${keyPrefix}/ssm-to-env-lambda-layer-example.zip`,
			),
			description:
				'This layer is used to pull config from SSM and convert to environmental variables',
		});

		new GuLambdaFunction(this, 'ssm-to-env-lambda-example', {
			app: 'ssm-to-env-lambda-example',
			fileName: 'ssm-to-env-lambda-example.zip',
			runtime: Runtime.NODEJS_16_X,
			handler: 'index.handler',
			environment: {
				AWS_LAMBDA_EXEC_WRAPPER: '/opt/ssm-to-env.sh',
			},
			layers: [getSecretsLayer],
		});
	}
}
