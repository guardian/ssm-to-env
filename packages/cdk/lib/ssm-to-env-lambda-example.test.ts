import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { SsmToEnvLambdaExample } from './ssm-to-env-lambda-example';

describe('The SsmToEnvLambdaExample stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new SsmToEnvLambdaExample(app, 'SsmToEnvLambdaExample', {
			stack: 'playground',
			stage: 'TEST',
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
