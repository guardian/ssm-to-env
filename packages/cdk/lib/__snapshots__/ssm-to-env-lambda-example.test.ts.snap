// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The SsmToEnvLambdaExample stack matches the snapshot 1`] = `
Object {
  "Parameters": Object {
    "DistributionBucketName": Object {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "getsecretslayer63FFACD4": Object {
      "Properties": Object {
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "playground/TEST/ssm-to-env-lambda-example/ssm-to-env-lambda-layer-example.zip",
        },
        "Description": "This layer is used to pull config from SSM and convert to environmental variables",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "ssmtoenvlambdaexample80E45003": Object {
      "DependsOn": Array [
        "ssmtoenvlambdaexampleServiceRoleDefaultPolicy4EA1D11C",
        "ssmtoenvlambdaexampleServiceRoleC035002F",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "playground/TEST/ssm-to-env-lambda-example/ssm-to-env-lambda-example.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "APP": "ssm-to-env-lambda-example",
            "AWS_LAMBDA_EXEC_WRAPPER": "/opt/ssm-to-env.sh",
            "STACK": "playground",
            "STAGE": "TEST",
          },
        },
        "Handler": "index.handler",
        "Layers": Array [
          Object {
            "Ref": "getsecretslayer63FFACD4",
          },
        ],
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "ssmtoenvlambdaexampleServiceRoleC035002F",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "ssm-to-env-lambda-example",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "47.0.0",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/ssm-to-env-lambda-example",
          },
          Object {
            "Key": "Stack",
            "Value": "playground",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ssmtoenvlambdaexampleServiceRoleC035002F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "ssm-to-env-lambda-example",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "47.0.0",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/ssm-to-env-lambda-example",
          },
          Object {
            "Key": "Stack",
            "Value": "playground",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ssmtoenvlambdaexampleServiceRoleDefaultPolicy4EA1D11C": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "DistributionBucketName",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/playground/ssm-to-env-lambda-example",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/playground/ssm-to-env-lambda-example/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ssmtoenvlambdaexampleServiceRoleDefaultPolicy4EA1D11C",
        "Roles": Array [
          Object {
            "Ref": "ssmtoenvlambdaexampleServiceRoleC035002F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
